File: bulk.js
import 'dotenv/config';
import { createClient } from '@supabase/supabase-js';
import XLSX from 'xlsx';
import fs from 'fs';
import path from 'path';
import fetch from 'node-fetch';

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

// -------------------- CONFIG --------------------
const FILE = './college.xlsx';
const PHOTO_DIR = './photos'; // local photo folder
const STAFF_SHEET = 'Sheet1';

const DAY_MAP = {
  M: 'monday',
  T: 'tuesday',
  W: 'wednesday',
  Th: 'thursday',
  F: 'friday',
  Sa: 'saturday',
};

function emptyWeek() {
  return {
    monday: Array(7).fill(null),
    tuesday: Array(7).fill(null),
    wednesday: Array(7).fill(null),
    thursday: Array(7).fill(null),
    friday: Array(7).fill(null),
    saturday: Array(7).fill(null),
  };
}
// -------------------- LOAD EXCEL --------------------
const workbook = XLSX.readFile(FILE);
const sheet = XLSX.utils.sheet_to_json(workbook.Sheets[STAFF_SHEET]);

if (!sheet.length) {
  throw new Error('Excel sheet empty or wrong sheet name');
}

// -------------------- HELPERS --------------------
async function uploadPhoto(photoValue, staffId) {
  if (!photoValue) return null;

  let buffer;
  let ext = 'jpg';

  // CASE 1: URL
  if (photoValue.startsWith('http')) {
    const res = await fetch(photoValue);
    if (!res.ok) throw new Error('Failed to fetch photo URL');
    buffer = Buffer.from(await res.arrayBuffer());
    ext = photoValue.split('.').pop().split('?')[0];
  }

  // CASE 2: Local file
  else {
    const filePath = path.join(PHOTO_DIR, photoValue);
    if (!fs.existsSync(filePath)) {
      throw new Error(`Photo file not found: ${photoValue}`);
    }
    buffer = fs.readFileSync(filePath);
    ext = path.extname(photoValue).replace('.', '');
  }

  const storagePath = `${staffId}/avatar.${ext}`;

  const { error } = await supabase.storage
    .from('avatars')
    .upload(storagePath, buffer, {
      upsert: true,
      contentType: `image/${ext}`,
    });

  if (error) throw error;

  const { data } = supabase.storage
    .from('avatars')
    .getPublicUrl(storagePath);

  return data.publicUrl;
}

// -------------------- MAIN --------------------
async function run() {
  for (const row of sheet) {
    const email = row['Mail id']?.trim();
    const name = row['Staff Name']?.trim();
    const photo = row['Photo']; // NEW COLUMN

    if (!email || !name) {
      console.warn('⏭ Skipping row (missing email/name)');
      continue;
    }

    console.log(`\n▶ Processing ${email}`);

    // -----------------------------------
    // 1. CREATE OR FETCH AUTH USER
    // -----------------------------------
    let userId;

    const { data: existing } = await supabase.auth.admin.listUsers();
    const found = existing.users.find(u => u.email === email);

    if (found) {
      userId = found.id;
      console.log('ℹ Auth user exists');
    } else {
      const { data, error } = await supabase.auth.admin.createUser({
        email,
        email_confirm: true,
        user_metadata: { full_name: name },
      });

      if (error) {
        console.error('❌ Auth error:', error.message);
        continue;
      }

      userId = data.user.id;
      console.log('✔ Auth user created');
    }

    // -----------------------------------
    // 2. WAIT FOR STAFF TRIGGER
    // -----------------------------------
    let staffId = null;
    for (let i = 0; i < 10; i++) {
      const { data } = await supabase
        .from('staff')
        .select('id')
        .eq('profile_id', userId)
        .maybeSingle();

      if (data?.id) {
        staffId = data.id;
        break;
      }
      await new Promise(r => setTimeout(r, 500));
    }

    if (!staffId) {
      console.error('❌ Staff row not found for', email);
      continue;
    }

    // -----------------------------------
    // 3. UPLOAD PHOTO (NEW)
    // -----------------------------------
    try {
      const photoUrl = await uploadPhoto(photo, staffId);
      if (photoUrl) {
        await supabase
          .from('staff')
          .update({ photo_url: photoUrl })
          .eq('id', staffId);

        console.log('✔ Photo uploaded');
      }
    } catch (e) {
      console.warn(`⚠ Photo skipped: ${e.message}`);
    }

    // -----------------------------------
    // 4. BUILD TIMETABLE
    // -----------------------------------
    const week = emptyWeek();

    for (const key of Object.keys(row)) {
      const m = key.match(/^(M|T|W|Th|F|Sa)(\d)$/);
      if (!m) continue;

      const day = DAY_MAP[m[1]];
      const idx = Number(m[2]) - 1;

      if (row[key]) {
        week[day][idx] = row[key].toString().trim();
      }
    }

    // -----------------------------------
    // 5. UPSERT TIMETABLE
    // -----------------------------------
    const { error: ttErr } = await supabase
      .from('timetable')
      .upsert(
        { staff_id: staffId, ...week },
        { onConflict: 'staff_id' }
      );

    if (ttErr) {
      console.error('❌ Timetable error:', ttErr.message);
      continue;
    }

    console.log('✔ Timetable saved');
  }

  console.log('\n🎉 IMPORT COMPLETED SUCCESSFULLY');
}

run().catch(err => {
  console.error('🔥 FATAL:', err);
});

File: bulk.py
import os
import re
import io
import time
import pandas as pd
from PIL import Image
from supabase import create_client
from dotenv import load_dotenv
from openpyxl import load_workbook

# ================= LOAD ENV =================
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")

supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)

# ================= CONFIG =================
FILE = "cse.xlsx"
BUCKET = "avatars"
DEPT = "CSE"
WEBP_QUALITY = 80  # Compression quality (1-100)

DAY_MAP = {
    "M": "monday",
    "T": "tuesday",
    "W": "wednesday",
    "Th": "thursday",
    "F": "friday",
    "Sa": "saturday",
}

# Creates 8 slots per day (indices 0 to 7)
def empty_week():
    return {day: [None] * 8 for day in DAY_MAP.values()}

# ================= IMAGE COMPRESSION HELPER =================
def compress_to_webp(image_bytes, quality=WEBP_QUALITY):
    """Converts raw image bytes to WebP format in-memory."""
    with Image.open(io.BytesIO(image_bytes)) as img:
        # Convert palette/CMYK modes to RGB for WebP compatibility
        if img.mode in ("P", "CMYK"):
            img = img.convert("RGB")
        
        output = io.BytesIO()
        img.save(output, format="WEBP", quality=quality, method=4)
        return output.getvalue()

# ================= LOAD AUTH USERS =================
print("📦 Loading auth users...")
auth_res = supabase.auth.admin.list_users(per_page=1000)
auth_users = {u.email: u for u in auth_res}

# ================= EXTRACT IMAGES =================
print("🖼 Extracting images from Excel...")
wb = load_workbook(FILE)
ws = wb.active

row_images = {}
for img in ws._images:
    excel_row = img.anchor._from.row + 1
    row_images[excel_row] = img._data()

# ================= LOAD EXCEL DATA =================
df = pd.read_excel(FILE)

# ================= MAIN LOOP =================
for index, row in df.iterrows():
    excel_row = index + 2  # header offset
    email = str(row.get("Mail id", "")).strip()
    name = str(row.get("Staff Name", "")).strip()

    if not email or email.lower() == "nan":
        continue

    print(f"\n▶ Processing {email}")

    # ---------- AUTH (CREATE OR UPDATE) ----------
    user = auth_users.get(email)

    if not user:
        res = supabase.auth.admin.create_user({
            "email": email,
            "email_confirm": True,
            "user_metadata": {"full_name": name},
        })
        user = res.user
        auth_users[email] = user
        print("👤 Auth user created")
    else:
        supabase.auth.admin.update_user_by_id(
            user.id,
            {"user_metadata": {"full_name": name}},
        )
        print("🔄 Auth user exists → metadata updated")

    user_id = user.id

    # ---------- STAFF UPSERT ----------
    staff_payload = {
        "profile_id": user_id,
        "dept": DEPT,
    }

    # ---------- IMAGE: COMPRESS → DELETE → UPLOAD ----------
    if excel_row in row_images:
        file_path = f"{user_id}/profile.webp"

        # 1. Convert raw bytes to compressed WebP
        webp_bytes = compress_to_webp(row_images[excel_row], quality=WEBP_QUALITY)

        # 2. Clean up both legacy JPG and existing WebP files
        try:
            supabase.storage.from_(BUCKET).remove([
                f"{user_id}/profile.jpg",
                f"{user_id}/profile.webp"
            ])
            print("🗑 Old photos removed")
        except Exception:
            pass

        # 3. Upload with upsert forced to true
        supabase.storage.from_(BUCKET).upload(
            file_path,
            webp_bytes,
            file_options={
                "content-type": "image/webp",
                "upsert": "true",
            },
        )

        # 4. Append cache-buster timestamp so frontend re-downloads fresh image
        timestamp = int(time.time())
        staff_payload["photo_url"] = (
            f"{SUPABASE_URL}/storage/v1/object/public/{BUCKET}/{file_path}?v={timestamp}"
        )

        print("✔ Photo compressed to WebP & uploaded")
    else:
        print(f"⚠ No image detected at Excel row {excel_row}")

    supabase.table("staff").upsert(
        staff_payload,
        on_conflict="profile_id",
    ).execute()

    # Fetch staff_id safely
    staff_row = (
        supabase.table("staff")
        .select("id")
        .eq("profile_id", user_id)
        .single()
        .execute()
    )

    staff_id = staff_row.data["id"]
    print("✔ Staff upserted")

    # ---------- TIMETABLE UPSERT ----------
    week = empty_week()

    for col in df.columns:
        match = re.match(r"^(M|T|W|Th|F|Sa)(\d)$", str(col))
        if match:
            day = DAY_MAP[match.group(1)]
            period_num = int(match.group(2))  # Period 1 to 7 from Excel
            value = row.get(col)

            if pd.notna(value) and str(value).strip():
                val_str = str(value).strip()

                if period_num in [1, 2, 3]:
                    # Periods 1, 2, 3 -> Indexes 0, 1, 2
                    week[day][period_num - 1] = val_str
                
                elif period_num == 4:
                    # FRIDAY EXCEPTION: No split lunch on Friday — everyone stays at index 3 (11:40 AM slot)
                    if day == "friday" or not re.search(r'\bS[12]\b', val_str.upper()):
                        week[day][3] = val_str  # Index 3 (Slot 4 -> 11:40 - 12:30)
                    else:
                        week[day][4] = val_str  # Index 4 (Slot 5 -> 12:30 - 1:20 for S1/S2)
                
                elif period_num in [5, 6, 7]:
                    # Periods 5, 6, 7 -> Indexes 5, 6, 7 (6th, 7th, 8th slots)
                    week[day][period_num] = val_str

    supabase.table("timetable").upsert(
        {"staff_id": staff_id, **week},
        on_conflict="staff_id",
    ).execute()

    print("✔ Timetable upserted")

print("\n🎉 IMPORT COMPLETED — SAFE TO RE-RUN")
